### PR TITLE
Keyboard editing improvements

### PR DESCRIPTION
**Pull request by titovmx**

### CODE REVIEW

File: src/themes/material/templates/body.cell.reference.edit.js [hunk 1]
```diff
@@ -30,25 +30,26 @@ export default function ReferenceEdit($scope, qgrid) {
...
-		this.cell().cancel.execute($cell, $event);
-		close();
+	this.commands = {
+		commit: new Command({
+			shortcut: 'ctrl+s',
+			execute: ($cell, $event) => {
+				this.cell().value = this.gridModel.selection().items;
+				this.cell().tag = {
+					entries: this.gridModel.selection().entries,
+					schema: this.gridModel.data().columns
```
Comment (vkorolev) [inline src/themes/material/templates/body.cell.reference.edit.js]: Should be `columns` instead of `schema`

Comment (titovmx) [reply src/themes/material/templates/body.cell.reference.edit.js]: Nice catch, thank you! 
Seems that auto merge hadn't seen code changes conflict here

File: src/core/infrastructure/command.manager.js [hunk 2]
```diff
@@ -3,6 +3,10 @@ export class CommandManager {
 		this.apply = apply;
 	}
 
+	canExecute() {
```
Comment (klumba12) [inline src/core/infrastructure/command.manager.js]: Key down should be here too

File: src/core/infrastructure/table.command.manager.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+export class TableCommandManager extends CommandManager {
+	constructor(apply, table) {
+		super(apply);
+
+		this.table = table;
+	}
+
+	keyDown(f) {
+		return this.table.view.keyDown(f);
```
Comment (klumba12) [inline src/core/infrastructure/table.command.manager.js:11]: Maybe make it as event for unification?

File: src/plugins/popup/popup.service.js [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
 import {AppError} from '@grid/core/infrastructure';
+import {noop} from '@grid/core/utility';
 
 export default class PopupService {
 	constructor() {
 		Array.from(arguments).forEach((arg, index) => this[PopupService.$inject[index]] = arg);
 
 		this.template = new TemplateLink(this.$compile, this.$templateCache);
 		this.popups = {};
+		this.shortcutOff = noop;
```
Comment (klumba12) [inline src/plugins/popup/popup.service.js]: why we need it here?

File: src/plugins/popup/popup.manager.js [hunk 5]
```diff
@@ -73,11 +77,27 @@ export default class Popup {
...
+
 	resize(settings) {
 		this.element.css({
 			width: Math.min(settings.width, this.body.clientWidth - this.element.offsetLeft) + 'px',
 			height: Math.min(settings.height, this.body.clientHeight - this.element.offsetTop) + 'px'
 		});
 		this.event.emit('resize');
 	}
+
+	commands(commands) {
```
Comment (klumba12) [inline src/plugins/popup/popup.manager.js]: what if the method'll be called sceral times? shortcutOff will be invalid, need to refactor this

File: src/themes/material/templates/body.cell.reference.edit.html [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <q-grid:popup id="q-grid-reference-popup" ng-init="$popup.show()"
 				  q-grid:popup-close="$view.edit.cell.cancel.execute($cell, $event)">
-	<q-grid:template for="trigger">
+	<q-grid:template for="trigger" let="[$ctrl]">
```
Comment (klumba12) [inline src/themes/material/templates/body.cell.reference.edit.html]: $ctrl is a bad name here

Comment (klumba12) [reply src/themes/material/templates/body.cell.reference.edit.html]: make it more meaningfull

File: src/themes/material/templates/body.cell.reference.edit.html [hunk 7]
```diff
@@ -22,15 +22,15 @@
 				</q-grid>
 			</div>
 			<div flex="10" class="q-grid-actions">
-				<md-button class="md-primary md-confirm-button" ng-click="$ctrl.commit($cell, $event)">Save
+				<md-button class="md-primary md-confirm-button" ng-click="$ctrl.commands.commit.execute($cell, $event)">Save
```
Comment (klumba12) [inline src/themes/material/templates/body.cell.reference.edit.html]: should be $ctrl.commit not $ctrl.commands.commit(unify)

File: src/plugins/edit-form/edit.form.panel.js [hunk 8]
```diff
@@ -1,33 +1,56 @@
...
+
+		this.reset = this.commands.reset;
+
+		this.shortcutOff = noop;
 	}
 
 	onInit() {
 		this.editor = new RowEditor(this.row, this.model.data().columns);
+
+		this.shortcutOff = this.shortcut.register('editFormManagement', new Map(
```
Comment (klumba12) [inline src/plugins/edit-form/edit.form.panel.js]: what if shortcut is not passed?

File: src/plugins/popup/popup.service.js [hunk 9]
```diff
@@ -43,6 +43,7 @@ export default class PopupService {
 		popupScope.id = settings.id;
 
 		const popup = angular.element('<q-grid:popup-panel id="id" model="model"></q-grid:popup-panel>'); // eslint-disable-line no-undef
+		this.popups[settings.id] = new PopupManager(popup, settings, this.$document[0].body);
```
Comment (klumba12) [inline src/plugins/popup/popup.service.js]: rename popupmanager to PopupEntry

File: src/themes/material/templates/body.cell.reference.edit.js [hunk 10]
```diff
@@ -1,15 +1,15 @@
...
 import {SelectionService} from '@grid/core/selection';
 
-ReferenceEdit.$inject = ['$scope', 'qgrid'];
-export default function ReferenceEdit($scope, qgrid) {
+ReferenceEdit.$inject = ['$scope', 'qgrid', 'qGridPopupService'];
+export default function ReferenceEdit($scope, qgrid, popupService) {
 	this.cell = () => $scope.$editor || $scope.$view.edit.cell;
 
-	let closed = false;
+	const id = 'q-grid-reference-popup';
```
Comment (klumba12) [inline src/themes/material/templates/body.cell.reference.edit.js]: ...reference-edit,,,

File: src/themes/material/templates/body.cell.reference.edit.js [hunk 11]
```diff
@@ -19,39 +19,63 @@ export default function ReferenceEdit($scope, qgrid) {
...
+	};
+
 	const dataChangedOff = this.gridModel.dataChanged.watch(e => {
 		if (e.hasChanges('rows') && e.state.rows.length > 0) {
 			const cell = this.cell();
 			this.gridModel.selection({
 				items: isArray(cell.value) ? cell.value : [cell.value]
 			});
 
+			this.gridModel.selectionChanged.watch(watchSelection);
```
Comment (klumba12) [inline src/themes/material/templates/body.cell.reference.edit.js:44]: register this outside of dataChanged

**Review by klumba12 — APPROVED** [review]

**Pull request merged**